### PR TITLE
Add guidelines on PR + commit text

### PR DESCRIPTION
**Pull request by bparees**
also create a guidelines directory and move the enhancement template
into it.

### CODE REVIEW

File: guidelines/commit_and_pr_text.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+* What changes are being made?  
+* Reference to any prior work(other PRs) or follow up work that will be needed.
+
+[Canonical example](https://github.com/openshift/openshift-apiserver/pull/69)
+
+# Git Commit Best practices
+
+## Number of Commits
+
+Strive for a single commit per PR.
```
Comment (alvaroaleman) [inline guidelines/commit_and_pr_text.md:29]: If we have this guideline - Why don't we simply make squash merging the global default (with the option to override via  a label, as we already offer today)?

Comment (bparees) [reply guidelines/commit_and_pr_text.md:29]: i don't think we want random commit text squashed together from 15 different commits.  i'd prefer people actually think through their commit messages and commits.

Comment (alvaroaleman) [reply guidelines/commit_and_pr_text.md:29]: Well, the commit body you get from github on squash merge is the PR description, then a list of commits with their title and body (if any). So the `random commit text` will be at the very end of the body and can be ignored if you don't care.

Squashmerging has the advantage that it allows ppl to do small and granular commits (which can be super useful during development) without cluttering the repos history.

Using the github PR description as body rather than a single commits body has IMHO a coupe of advantages as well:
* For ppl that already write commit messages, nothing changes, because github will put the body into the pr description if its only one commit
* Ppl that are used to write proper descriptions in Github but are not used to write commit bodies do not have to change their workflow
* Reviewing the PR description is much easier than reviewing the commit body (You don't have to select the commit and yalayala/check  the pr out locally).

File: guidelines/commit_and_pr_text.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+Minimal commits per PR make it easier to review the PR, easier to revert the PR if necessary, easier for someone doing “git blame” to follow the set of involved changes, and encourage smaller “chunks of work” (if it’s multiple commits, think about whether it should also be multiple PRs)
```
Comment (jmrodri) [inline guidelines/commit_and_pr_text.md:31]: One thing I try to do is make my vendor commits separate from all other work in the PR. I find it incredibly difficult to review 14,000 vendor changes :)

File: guidelines/commit_and_pr_text.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# Motivation
+Our changelogs (reviewed by users/customers) are autogenerated from PR titles and git commit text.  The better this information is, the more we help our customers.  This information also helps reviewers understand changes, patch managers understand risk/importance, and when hunting for regressions, helps us better understand what is/isn’t likely to be a cause.
```
Comment (jmrodri) [inline guidelines/commit_and_pr_text.md:2]: On the SDK team we recently went through this process of changelog. We used to have a single changelog file that would need to be edited for EACH PR. This caused a bunch of subsequent PRs to have merge conflicts as the file was updated.

We also thought about making the commit messages the source of truth. This proved to be quite annoying as it was a departure from our current workflow, meant much more diligence to the commit messages.

Our final solution was to create the changelog from a collection of fragments that were added to each PR. This allowed us to better describe the change for the customer/user in the changelog.

https://github.com/operator-framework/operator-sdk/pull/2761

File: guidelines/commit_and_pr_text.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+* Do not let your PR title get truncated/wrapped into the description
+* Additional prefixes like what release the PR applies to are allowed, but they will be stripped from the changelog (the changelog for 4.4 doesn’t need to say “release-4.4” 
+  * Changelog strips anything between square brackets that has no spaces - i.e. [release-4.4] is stripped but [release 4.4] would not be
+    
+
+__PR description:__  Additional context reviewers need to understand your PR
+
+* Why was this change needed?
+* What changes are being made?  
+* Reference to any prior work(other PRs) or follow up work that will be needed.
```
Comment (jmrodri) [inline guidelines/commit_and_pr_text.md:21]: In general, I try to make my commits as meaningful as possible. Often like an email where the first line is the Subject. The description is more details about what is going on in the PR, motivations links to other related works, etc.

File: guidelines/commit_and_pr_text.md (hunk 1 above)
Comment (jmrodri) [reply guidelines/commit_and_pr_text.md:29]: I prefer squash commits too. I've configured many of my repos to be squash commit.

File: guidelines/commit_and_pr_text.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+Strive for a single commit per PR.
+
+Minimal commits per PR make it easier to review the PR, easier to revert the PR if necessary, easier for someone doing “git blame” to follow the set of involved changes, and encourage smaller “chunks of work” (if it’s multiple commits, think about whether it should also be multiple PRs)
+
+Valid reasons to have multiple commits:
+
+* Separating implementation work from generated code or vendor updates
+* Separating api work from implementation
+* Moving files around in a refactor (do the file move in one commit, edit the files in the second commit, so that the edits stand out in the commit diff)
+* Temporarily adding additional commits in response to reviewer feedback so reviewers can easily see how their feedback was addressed.  These commits should be squashed prior to merging.
```
Comment (jmrodri) [inline guidelines/commit_and_pr_text.md:38]: Okay thanks for calling this out. This makes sense to me.

File: guidelines/commit_and_pr_text.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+Invalid reasons to have multiple commits
+* Separating tests+implementation
+* Committing+pushing your work each day/hour (use amend instead, or at least ensure you squash before merging).
```
Comment (jmrodri) [inline guidelines/commit_and_pr_text.md:42]: I often do this especially to address review comments.

File: guidelines/commit_and_pr_text.md (hunk 1 above)
Comment (mrunalp) [reply guidelines/commit_and_pr_text.md:29]: On the other hand, splitting your PR into commits make them easier to review in chunks. Also, we should add text around making sure that the commits can be bisectable i.e. tests should pass at each commit.

File: guidelines/commit_and_pr_text.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+## Git Commit text
+
+__First line:__  An english language sentence with capitalization, no period, max 50 characters
```
Comment (mrunalp) [inline guidelines/commit_and_pr_text.md:46]: We can add a guideline for wrapping the commit message text at 72 characters with exceptions for links or error messages or quotes.

**Comment by bparees** [comment]
> I think I'm okay with these guidelines. Will they be enforced at all? If not, then +1 if we are planning on having automation I'd be against that, often that turns into a 'how can I appease the bot so I can get on with my work' scenario.

no plans to enforce, just tips for the org to follow and something to point to when a reviewer sees a PR/commit that doesn't seem sufficient.

File: guidelines/commit_and_pr_text.md (hunk 2 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:31]: covered by https://github.com/openshift/enhancements/pull/311/files#diff-8092445902fc4142135544e8ba663eb1R35 no?

File: guidelines/commit_and_pr_text.md (hunk 3 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:2]: the changelog isn't a committed file in the repo, it's auto generated doc that's available w/ the release.

File: guidelines/commit_and_pr_text.md (hunk 4 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:21]: then you don't need this doc because you're doing it right already :)

File: guidelines/commit_and_pr_text.md (hunk 6 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:42]: yeah so that's fine to do pre-merge, but a PR shouldn't be LGTMed while still in that state.

File: guidelines/commit_and_pr_text.md (hunk 7 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:46]: links probably just don't belong in the first line of a commit message.  nor error messages or quotes.

Comment (mrunalp) [reply guidelines/commit_and_pr_text.md:46]: I meant we add that guideline for the commit message body.

Comment (bparees) [reply guidelines/commit_and_pr_text.md:46]: @mrunalp added

Comment (mrunalp) [reply guidelines/commit_and_pr_text.md:46]: Thanks!

File: guidelines/commit_and_pr_text.md (hunk 3 above)
Comment (mrunalp) [reply guidelines/commit_and_pr_text.md:2]: I think that the kubernetes PR template https://github.com/kubernetes/kubernetes/edit/master/.github/PULL_REQUEST_TEMPLATE.md (a form of which CRI-O has adopted https://github.com/cri-o/cri-o/blob/master/.github/PULL_REQUEST_TEMPLATE.md) works well for generating changelog. It includes a `release-note` section that can be parsed to generate a changelog. It helps when a PR has multiple commits. The whole template itself is very useful to a reviewer for understanding the motivation for a PR.

File: guidelines/commit_and_pr_text.md (hunk 1 above)
Comment (markmc) [reply guidelines/commit_and_pr_text.md:29]: Squashing is good for avoiding lots of fixup commits in the history, but it is perfectly reasonable for a PR to have a well-structured series of commits (with helpful commit messages) where each commit is a separate logical change.

Comment (bparees) [reply guidelines/commit_and_pr_text.md:29]: For everyone commenting on this goal....It says "strive for" not "thou shalt", and is followed by a section titled "valid reasons to have multiple commits".   This is not saying we can never have multiple commits.  If you think there are additional reasons not covered by the existing list, please add suggestions to that list.  If this is a phrasing issue i'm happy to accept suggestions on a different wording that makes it clear...  the intent is "you should generally default to one commit per PR unless you fall into one of the exception categories"

I want to emphasize that in general these are guidelines not rules.   No one is going to be enforcing them strictly, we're just trying to get back to a little more discipline on our PR/git hygiene and give guidance to people who may be less experienced in what a good pr/commit looks like.

File: guidelines/commit_and_pr_text.md (hunk 2 above)
Comment (cfergeau) [reply guidelines/commit_and_pr_text.md:31]: > Minimal commits per PR make it easier to review the PR, [...], easier for someone doing “git blame” to follow the set of involved changes, and encourage smaller “chunks of work” (if it’s multiple commits, think about whether it should also be multiple PRs)

When I read this, I can't help but think there are 2 conflicting things here,
1) minimal commits
2) minimizing the number of commits per PR

1) indeed helps with reviews, git blame, ... 2) conflicts with that in that it will not encourage people to split their work meaningfully since the recommendations push to use only 1 commit per PR.
In my experience, the smaller a commit is, the easier it is to review. If you use `git blame` because you want to understand why a change was made, the smaller the commit is, the more specific the commit message will be as to why the change was made. If you use `git bisect` to track down a bug on code you are not much familiar with, you also want to end up on a commit making as little changes as possible. Except for trivial changes/bug fixes, it's usually possible to split a change in several meaning full commits.
Actually, looking at the commit log of the 'canonical example' below ( https://github.com/openshift/origin/commit/d7a7ff790f87ae14ad106f2256ebafc1e7c91949 ), I'm under the impression each paragraph could have deserved its own commit, and this would have made it easier to assess (for example) whether "Remove the 'break' from the wait-loop's ClusterOperator iteration. The previous logic would only list the first unavailable operator. With this commit we now list each unavailable operator while we wait." does not introduce unwanted/unexpected logic changes. As it is now, the change is drowned in other unrelated changes, making the review process harder, and making a potential revert of just this part harder too.

File: guidelines/commit_and_pr_text.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+__Additional information:__  Why was this change needed?  What is the gist of the change?
+
+* Wrap lines at 72 characters, barring exceptions for links or other content that can't be wrapped.
+* Avoid referencing github issues/PRs if you are repeatedly amending the commit as it generates a lot of noise in the referenced issue/PR.
+
```
Comment (cfergeau) [inline guidelines/commit_and_pr_text.md:52]: I don't know if it's worth insisting that the "Additional information:" part is very important, and not optional? I've seen too many one line commit logs in openshift repositories which are quite cryptic for the occasional reader of the code :(

github issue references in logs are very useful when using `git blame` or similar. I've been told that instead of amending a commit/force pushing, one can use `--fixup` and that github will squash all the fixups at merge time regardless of the merge settings on the branch (I haven't tried it yet...)

File: guidelines/commit_and_pr_text.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
+# Motivation
+Our changelogs (reviewed by users/customers) are autogenerated from PR titles and git commit text.  The better this information is, the more we help our customers.  This information also helps reviewers understand changes, patch managers understand risk/importance, and when hunting for regressions, helps us better understand what is/isn’t likely to be a cause.
+
+# PR Metadata
+
+__PR title:__ Bug NNNN: A short sentence with capitalization and no period or the Jira tag for the issue “Feature TEAM-XXX:”
```
Comment (lilic) [inline guidelines/commit_and_pr_text.md:6]: Problem for monitoring team is we have a lot of dependancies that are bumped with one single PR, when updating jsonnet deps, so how would we describe this in short? We now say something vague like "Bump dependencies". What would you suggest for us to write here instead?

File: guidelines/commit_and_pr_text.md (hunk 8 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:52]: > I don't know if it's worth insisting that the "Additional information:" part is very important, and not optional? 

I can reword to make it a stronger request/suggestion, but I'd like to avoid absolute requirements/rules since:

1) right now we have no plans to enforce anyway
2) "there's always an exception"

File: guidelines/commit_and_pr_text.md (hunk 2 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:31]: I'm changing the text to `Minimize your number of commits per PR with a goal of a single commit in many cases.`

(again with the reasons and exceptions listed below that text).  

> The 1st item indeed helps with reviews, git blame, ... 2) conflicts with that in that it will not encourage people to split their work meaningfully since the recommendations push to use only 1 commit per PR.

Minimal *number* of commits helps with PR review imho, mainly because if you've got a PR with a vendor commit and then 50 other commits, I can't easily exclude the vendor commit from my review (I can either view all changes, or i have to go through each commit by itself).  With a vendor commit plus one or two commits, I can easily review the one or two commits that aren't the vendor commit.  That was the genesis of the statement that "minimal commits per PR make it easier to review the PR".

Certainly to some extent this probably depends on the nature of the PR as well as how reviewers prefer to operate.

Minimal sized commits (which weren't really discussed in this doc, i will rephrase "minimal commits to minimal numbers of commits, in the text) wasn't intended as a goal of this doc.  Commits should be as big or as small as they need to be do to the related/cohesive task they are intended for.  Sometimes that's 1 line in 1 file and sometimes that's 1000 lines in 1000 files (think utility function refactoring).

> 2) conflicts with that in that it will not encourage people to split their work meaningfully since the recommendations push to use only 1 commit per PR.

I think this is where the "consider if it should have been multiple PRs" statement comes in.


> Actually, looking at the commit log of the 'canonical example' below ( openshift/origin@d7a7ff7 ), I'm under the impression each paragraph could have deserved its own commit, 

point taken, the intent was that this was a good example of commit text (explaining what was done), not necessarily a good example of how to break down a body of work into specific commits.  I will make that clear.

File: guidelines/commit_and_pr_text.md (hunk 9 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:6]: I think that's a totally reasonable summary sentence for a dep bump.

however if there's anything special worthy of note about the bump, it can be put in the longer description (e.g. did you have a pin a dep version or add a new one?  did other code have to refactored in reaction to the bump, in which case those changes should probably be a separate commit in the same PR?)

**Comment by bparees** [comment]
Pushed a few tweaks (in a new commit that will be squashed before merge!) in response to some of the feedback so far.

File: guidelines/commit_and_pr_text.md (hunk 8 above)
Comment (cfergeau) [reply guidelines/commit_and_pr_text.md:53]: Yup, of course there are always exceptions :) `Fix 'requierd' typo in doc` does not really more than than :)

File: guidelines/commit_and_pr_text.md (hunk 2 above)
Comment (cfergeau) [reply guidelines/commit_and_pr_text.md]: > > Actually, looking at the commit log of the 'canonical example' below ( [openshift/origin@d7a7ff7](https://github.com/openshift/origin/commit/d7a7ff7) ), I'm under the impression each paragraph could have deserved its own commit,
> 
> point taken, the intent was that this was a good example of commit text (explaining what was done), not necessarily a good example of how to break down a body of work into specific commits. I will make that clear.

For what it's worth, it was clear that it was only a specific example in the context of 'good commit text', it was just convenient to reuse this one as a counter example.


> Minimal _number_ of commits helps with PR review imho, mainly because if you've got a PR with a vendor commit and then 50 other commits, I can't easily exclude the vendor commit from my review (I can either view all changes, or i have to go through each commit by itself). With a vendor commit plus one or two commits, I can easily review the one or two commits that aren't the vendor commit. That was the genesis of the statement that "minimal commits per PR make it easier to review the PR".
> 
> Certainly to some extent this probably depends on the nature of the PR as well as how reviewers prefer to operate.

Hmm it seems we definitely operate very differently, I have a hard time reviewing efficiently/correctly commits doing multiple things which could have been separated in smaller logical commits :-/ But maybe your "50 commits" example would be 50 commits such as "adjust previous commit", "fix bug in 5th commit of PR", "remove debug code added in PR", ...? In which case, yes, I agree it would be complicated to review.

> Minimal sized commits (which weren't really discussed in this doc, i will rephrase "minimal commits to minimal numbers of commits, in the text) wasn't intended as a goal of this doc.

Which I'm worried is going to be a disincentive to splitting commits. As the developer, I'd just read that `git commit -a` and then PR is recommended by this doc :) This imo makes the reviewers task more complicated.

Comment (bparees) [reply guidelines/commit_and_pr_text.md]: > But maybe your "50 commits" example would be 50 commits such as "adjust previous commit", "fix bug in 5th commit of PR", "remove debug code added in PR", ...? In which case, yes, I agree it would be complicated to review.

That is definitely the clearest example of the kind of commit behavior we want to avoid.

> Which I'm worried is going to be a disincentive to splitting commits. As the developer, I'd just read that git commit -a and then PR is recommended by this doc :) This imo makes the reviewers task more complicated.

well again, the idea here is that should be your default mode of operation(and that before you create that second commit, you consider whether it should instead be a second PR), but that there are valid reasons to have multiple commits.

File: guidelines/commit_and_pr_text.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
+# Motivation
+Our changelogs (reviewed by users/customers) are autogenerated from PR titles and git commit text.  The better this information is, the more we help our customers.  This information also helps reviewers understand changes, patch managers understand risk/importance, and when hunting for regressions, helps us better understand what is/isn’t likely to be a cause.
```
Comment (russellb) [inline guidelines/commit_and_pr_text.md:2]: I would add something about current and future team members trying to understand why things are they way they are, as well.  The commit log has the history of what we know about what was changed and why.

File: guidelines/commit_and_pr_text.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+* Try to describe the problem the user sees, rather than the code change you are doing
+  * I.e. “Console does not work when user clicks X” instead of “Fix drawer notification integration with underlying controller”
+* Prefix bug fixes with the bug number, “Bug XXXXX:” 
+  * (capitalize bug, space, number, then colon)
+* Do not let your PR title get truncated/wrapped into the description
+* Additional prefixes like what release the PR applies to are allowed, but they will be stripped from the changelog (the changelog for 4.4 doesn’t need to say “release-4.4” 
+  * Changelog strips anything between square brackets that has no spaces - i.e. [release-4.4] is stripped but [release 4.4] would not be
+    
+
+__PR description:__  Additional context reviewers need to understand your PR
```
Comment (russellb) [inline guidelines/commit_and_pr_text.md:17]: Maybe add:

 * All of this information must also reside in commit messages, and not only in the PR summary and description

File: guidelines/commit_and_pr_text.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+## Git Commit text
+
+__First line:__  An english language sentence with capitalization, no period, max 50 characters
+
+__Additional information:__  Why was this change needed?  What is the gist of the change?
+
+* Almost every commit should have more than just a single line descriptor.  If you are making a commit with no additional information, you should think twice about whether you've really provided enough information that another developer can understand the intent of the commit.
+* Wrap lines at 72 characters, barring exceptions for links or other content that can't be wrapped.
+* Avoid referencing github issues/PRs if you are repeatedly amending the commit as it generates a lot of noise in the referenced issue/PR.
+
```
Comment (russellb) [inline guidelines/commit_and_pr_text.md:53]: Some more ideas for this section:

 * If the commit fixes a bug, describe in detail the problem that was observed that led up to this change
 * Describe alternatives considered and why this change is the best one to make now, if applicable

File: guidelines/commit_and_pr_text.md (hunk 11 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:17]: I think that's covered by the git commit text section below.  I don't want to say that the two should be identical because of things like:

`Avoid referencing github issues/PRs if you are repeatedly amending the commit as it generates a lot of noise in the referenced issue/PR.`

File: guidelines/commit_and_pr_text.md (hunk 12 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:53]: added

File: guidelines/commit_and_pr_text.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+## Git Commit text
+
+__First line:__  An english language sentence with capitalization, no period, max 50 characters
+
+__Additional information:__  Why was this change needed?  What is the gist of the change?
+
+* Almost every commit should have more than just a single line descriptor.  If you are making a commit with no additional information, you should think twice about whether you've really provided enough information that another developer can understand the intent of the commit.
+* Wrap lines at 72 characters, barring exceptions for links or other content that can't be wrapped.
+* Avoid referencing github issues/PRs if you are repeatedly amending the commit as it generates a lot of noise in the referenced issue/PR.
```
Comment (russellb) [inline guidelines/commit_and_pr_text.md:52]: I find references incredibly valuable, but this caveat does seem valuable.  How about this:

 * Add all relevant references to the commit text, including other commits, bugs, PRs, or other links that help explain or understand the change.  Note however that you should minimize amending a commit with github references, as it will generate a lot of noise on reference issues or PRs.  Instead, consider adding commits while iterating on a PR and squashing them once before merging.

This also seems like "github best practices", maybe to be added up in the github section instead of here?

File: guidelines/commit_and_pr_text.md (hunk 10 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md]: added

File: guidelines/commit_and_pr_text.md (hunk 11 above)
Comment (russellb) [reply guidelines/commit_and_pr_text.md:22]: Right, I don't think they should be identical either.  What I don't like seeing is good detail in a PR summary but single-line unhelpful commit messages.

**Comment by russellb** [comment]
@bparees looks like you have some unrelated file renames in the PR

File: guidelines/commit_and_pr_text.md (hunk 13 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md]: revised

**Comment by bparees** [comment]
> @bparees looks like you have some unrelated file renames in the PR

it was intentional(some minor but long overdue cleanup in this repo) but in the interest of practicing what i preach i will remove them.

File: guidelines/commit_and_pr_text.md (hunk 9 above)
Comment (soltysh) [reply guidelines/commit_and_pr_text.md:11]: Maybe something like https://github.com/openshift/cluster-kube-controller-manager-operator/pull/406#issue-409539505, maybe not necessarily as detailed.

File: guidelines/commit_and_pr_text.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+* What changes are being made?  
+* Reference to any prior work(other PRs) or follow up work that will be needed.
+
+[Canonical PR text example](https://github.com/openshift/openshift-apiserver/pull/69)
+
+# Git Commit Best practices
+
+## Number of Commits
+
+Minimize your number of commits per PR with a goal of a single commit in many cases.
```
Comment (abhinavdahiya) [inline guidelines/commit_and_pr_text.md:34]: is it really the best indicator of better reviewable PR? if there are various commits with smaller logical changes, for such a PR is much easy to review instead of one commit with very large change.

personally I feel asking people to make one commit PRs would force people to put even large changes in one commit. 

Maybe instead of number of commit we talk about _non-vendor/non-generated_ changeset size?

File: guidelines/commit_and_pr_text.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+Minimal numbers commits per PR make it easier to review the PR, easier to revert the PR if necessary, easier for someone doing “git blame” to follow the set of involved changes, and encourage smaller “chunks of work” (if it’s multiple commits, think about whether it should also be multiple PRs)
```
Comment (abhinavdahiya) [inline guidelines/commit_and_pr_text.md:36]: we don't squash commits on PR merge, so it is usually very easy to go from commit to the PR even if there were multiple commits in the PR using the github UI or git cli, and multiple commits are really not that problematic for tracking as long as they are focused... in my experience.

File: guidelines/commit_and_pr_text.md (hunk 14 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:34]: > is it really the best indicator of better reviewable PR? 

It was not claimed as the best indicator of anything. it was positioned as a desirable attribute within the conditions/caveats discussed later.

> if there are various commits with smaller logical changes, for such a PR is much easy to review instead of one commit with very large change.

there was discussion about this elsewhere in this PR, some people review commit by commit, others do it at a PR level.  Which approach makes sense depends in part of how the PR is structured.

> personally I feel asking people to make one commit PRs would force people to put even large changes in one commit.

I want to emphasize this as much as i can:  there is no requirement to make single commit PRs.  The ask is to think carefully about how many commits you are putting in a PR and why.  Single commit is a goal(where applicable) and not a requirement.

Comment (cfergeau) [reply guidelines/commit_and_pr_text.md:34]: >  Single commit is a goal(where applicable)

This is the concern, single commit is strongly recommended, so why bother with splitting for ease of maintainance/review/...?

Comment (bparees) [reply guidelines/commit_and_pr_text.md:34]: > This is the concern, single commit is strongly recommended, so why bother with splitting for ease of maintainance/review/...?

i'm going to reframe this statement to try to alleviate the concerns people are raising about it.

**Comment by bparees** [comment]
@cfergeau @abhinavdahiya @markmc @alvaroaleman i've pushed a new commit rephrasing the "multiple commits in a pr" guidance.  ptal and see if it satisfies the concerns you've raised.

File: guidelines/commit_and_pr_text.md (hunk 9 above)
Comment (lilic) [reply guidelines/commit_and_pr_text.md:11]: Nice, that makes sense. Any objections @openshift/openshift-team-monitoring maybe we can just find a way to somehow copy over the commits from the upstream deps, e.g. kube-mixin commits?

File: guidelines/commit_and_pr_text.md (hunk 6 above)
Comment (ashcrow) [reply guidelines/commit_and_pr_text.md:45]: fixups + `git rebase -i [...]` is your friend when doing this on multiple separate commits. End result is squashed, targeted commits :+1:

File: guidelines/commit_and_pr_text.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+* Moving files around in a refactor (do the file move in one commit, edit the files in the second commit, so that the edits stand out in the commit diff)
+* Temporarily adding additional commits in response to reviewer feedback so reviewers can easily see how their feedback was addressed.  These commits should be squashed prior to merging.
+
+Invalid reasons to have separate commits
+* Separating tests+implementation
+* Committing+pushing your work each day/hour (use amend instead, or at least ensure you squash before merging).
+
+## Git Commit text
+
+__First line:__  An english language sentence with capitalization, no period, max 50 characters
```
Comment (ashcrow) [inline guidelines/commit_and_pr_text.md:49]: I would argue that area of modification is quite helpful. Here are some PR title examples from MCO:

- `cmd: Configure klog to match glog`: https://github.com/openshift/machine-config-operator/pull/1730
- `pkg/daemon: Add IgnitionVersion to Daemon`: https://github.com/openshift/machine-config-operator/pull/1729
- `mcs: Log User-Agent`: https://github.com/openshift/machine-config-operator/pull/1705

This is very helpful when going through history as well as when reviewing changes. Directly at a glance each PR's focus is quickly understood. On top of that, using this pattern helps folks avoid the all-the-things commits _unless_ it explicitly must be done (say -- refactoring variable names, etc.. though one could say the area would be `refactor:` :smile:) 

Please note, these examples do end up showing functionality rather than the problem a user sees. Combining "describe the problem" with "area of modification" would be quite a nice clarification overall!

File: guidelines/commit_and_pr_text.md [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+__Additional information:__  Why was this change needed?  What is the gist of the change?
+
+* Almost every commit should have more than just a single line descriptor.  If you are making a commit with no additional information, you should think twice about whether you've really provided enough information that another developer can understand the intent of the commit.
+* Wrap lines at 72 characters, barring exceptions for links or other content that can't be wrapped.
+* If the commit fixes a bug, describe in detail the problem that was observed that led up to this change
+* Describe alternatives considered and why this change is the best one to make now, if applicable
+* Add all relevant references to the commit text, including other commits, bugs, PRs, or other links that help explain or understand the change. Note however that you should minimize amending a commit with github references, as it will generate a lot of noise on reference issues or PRs. Instead, consider adding commits while iterating on a PR and squashing them once before merging.
+
```
Comment (ashcrow) [inline guidelines/commit_and_pr_text.md:58]: I've found tl;dr docs in the commit body to be helpful. Essentially the end result change for folks who are reviewing. For example, if `COMPONENTNAME` decided to change some environment variables the following might be _part of_ the body:

```
[...]

Environment variable renames:
- CONFIG_FILE -> COMPONENTNAME_CONFIG
- LOG_LEVEL -> COMPONENTNAME_LOG_LEVEL
- LOG_FILE -> COMPONENT_LOG_PATH
```

File: guidelines/commit_and_pr_text.md (hunk 16 above)
Comment (bparees) [reply guidelines/commit_and_pr_text.md:49]: Are you suggesting this for the PR title or the commit text?  Given the commit text ends up in the release changelog i think it's undesirable in the commit text (first line).

having it in the PR title, or in the git text body seems ok, though it very much feels like a "ymmv" thing for how useful it is for different teams so i'm a little hesitant to put it in here as anything more than a "something to consider if you like it" suggestion.

**Comment by bparees** [comment]
@ashcrow comments replied to and/or addressed in a new commit for you.

File: guidelines/commit_and_pr_text.md (hunk 16 above)
Comment (ashcrow) [reply guidelines/commit_and_pr_text.md:50]: commit text (first line). I can accept your reasoning against it though @bparees. Generally I think the area pattern is helpful, but if things are flowing into the changelog and could be a bit awkward it would be understandable to note specify the area of change that way. It may be helpful in the text body, _but_ at that level I'd agree it's more of a "something to consider if you like it" at best.

File: guidelines/commit_and_pr_text.md [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+__PR title:__ Bug NNNN: A short sentence with capitalization and no period or the Jira tag for the issue “Feature TEAM-XXX:”
+
+* Try to describe the problem the user sees, rather than the code change you are doing
+  * I.e. “Console does not work when user clicks X” instead of “Fix drawer notification integration with underlying controller”
+* Prefix bug fixes with the bug number, “Bug XXXXX:” 
+  * (capitalize bug, space, number, then colon)
+* Do not let your PR title get truncated/wrapped into the description
+* Additional prefixes like what release the PR applies to are allowed, but they will be stripped from the changelog (the changelog for 4.4 doesn’t need to say “release-4.4” 
+  * Changelog strips anything between square brackets that has no spaces - i.e. [release-4.4] is stripped but [release 4.4] would not be
```
Comment (sdodson) [inline guidelines/commit_and_pr_text.md:19]: Just wanted to confirm, which process actually does this as I only knew this to be the practice of patch managers but not an automated process?

File: guidelines/commit_and_pr_text.md (hunk 16 above)
Comment (derekwaynecarr) [reply guidelines/commit_and_pr_text.md:50]: the area pattern is heavily used in installer if i recall.

/cc @abhinavdahiya should give his perspective as well.

**Comment by bparees** [comment]
> are there any guidelines we want to enforce via tooling?

personally i'd rather not, there are always going to be "i just need to fix this typo" commits and i can't think of a minimal rule that's both:

1) going to allow people to easily fix minor issues w/o feeling frustrated about having to satisfy some bot requirement
2) going to provide value by enforcing some minimal standard/information content

to me this really is just an awareness thing that reviewers should help train PR submitters into doing properly.

**Comment by ashcrow** [comment]
I second @bparees notes above. This is a solid step and we can iterate with modifications, tooling, etc.. over time as appropriate.

**Comment by derekwaynecarr** [comment]
@bparees ok, given the above around enforcement, then this doc looks like great guidance.

/approve
/lgtm
